source-oracle/replication.go

```diff
@@ -0,0 +1,902 @@
+}
+
+// run is the main loop of the replicationStream which loops message
+// receiving and relaying
+func (s *replicationStream) run(ctx context.Context) error {
+	var poll = time.NewTicker(pollInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-poll.C:
+		}
+
+		var err = s.poll(ctx)
+		if err != nil {
+			// ORA-01013: user requested cancel of current operation
+			// this means a cancellation of context happened, which we don't consider an error
+			// since it should only happen in tests
+			if strings.Contains(err.Error(), "ORA-01013") {
+				continue
+			}
+			return fmt.Errorf("failed to poll messages: %w", err)
+		}
+	}
+}
+
+func (s *replicationStream) poll(ctx context.Context) error {
+	for {
+		switched, err := s.redoFileSwitched(ctx)
+		if err != nil {
+			return err
+		}
+
+		if switched {
+			if err := s.endLogminer(ctx); err != nil {
+				return err
+			} else if err := s.startLogminer(ctx, s.lastTxnEndSCN); err != nil {
+				return err
+			}
+		}
+
+		msgs, err := s.receiveMessages(ctx)
+		if err != nil {
+			return fmt.Errorf("receive messages: %w", err)
+		}
+
+		for _, msg := range msgs {
+			event, err := s.decodeMessage(msg)
+			if err != nil {
+				return fmt.Errorf("error decoding message: %w", err)
+			}
+
+			s.events <- event
+		}
+
+		s.events <- &sqlcapture.FlushEvent{
+			Cursor: strconv.Itoa(s.lastTxnEndSCN),
+		}
+	}
+}
+
+// decode the escaped unicode literals in UNISTR() functions
+// the unicode literals are in the UCS-2 format
+// see https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/UNISTR.html
+func decodeUnistr(input string) (string, error) {
+	var val string
+	for i := 0; i < len(input); i++ {
+		var c = input[i]
+		// 92 = \ (backslash)
+		if c == 92 {
+			// two backslashes following each other is just an escaped backslash
+			if len(input) > i+1 && input[i+1] == 92 {
+				val += string(c)
+				i += 1
+			} else if len(input) > i+4 {
+				var codeStrOne = fmt.Sprintf("%s", input[i+1:i+5])
+				var codeOne, err = strconv.ParseInt(codeStrOne, 16, 32)
+				if err != nil {
+					return "", fmt.Errorf("parsing unicode point at %q (%d): %w", input, i+1, err)
+				}
+				if utf16.IsSurrogate(rune(codeOne)) {
+					var codeStrTwo = fmt.Sprintf("%s", input[i+6:i+10])
+					var codeTwo, err = strconv.ParseInt(codeStrTwo, 16, 32)
+					if err != nil {
+						return "", fmt.Errorf("parsing unicode point at %q (%d): %w", input, i+6, err)
+					}
+					i += 5
+
+					val += string(utf16.DecodeRune(rune(codeOne), rune(codeTwo)))
+				} else {
+					val += string(rune(codeOne))
+				}
+
+				i += 4
+			} else {
+				val += string(c)
+			}
+		} else {
+			val += string(c)
+		}
+	}
+
+	return val, nil
+}
+
+// decode SQL values extracted from the sql queries returned by logminer and parsed by sqlparser, into values
+// to be used in the document
+func decodeValue(colInfo sqlcapture.ColumnInfo, expr sqlparser.Expr, colName string, originalSql string) (any, error) {
```

Comment:
`colInfo`, `colName`, and `originalSql` aren't used anywhere in this function - should they be? Or maybe the can be removed as inputs?